float32 startX
float32 startY
float32 endX
float32 endY
uint8 type